### ISSUE TITLE
whish for 2016: matplotlib can use only Pillow 3.0+ to create animated GIF

### ISSUE THREAD

**Issue by stonebig** [matplotlib/matplotlib#5750]
Imagemagick is a problem on windows, and the `animated gif` functionality of Pillow seems to work for some well-known situations since 3.0.0.

(would spare 40 Mo of Imagemagick downloads on Windows 64bit, and installation  tricks, a huge win)

**Comment by tacaswell** [comment]
Someone just needs to write a pillow-based writer.  This should be pretty simple once https://github.com/matplotlib/matplotlib/pull/5454 gets merged.

**Comment by WarrenWeckesser** [comment]
> the animated gif functionality of Pillow seems to work for some well-known situations since 3.0.0.

Do you have an up-to-date example that demonstrates this?   This would be a good use-case for https://github.com/matplotlib/matplotlib/pull/5454.  A working script that uses Pillow to programmatically create an animated GIF from a sequence of individual images (either Pillow `Image` instances or numpy arrays) would be a great help.  I did  a little googling, but it is hard to tell what would be the best practice when using Pillow 3 (e.g. the file `gifmaker.py` doesn't actually contain much code these days; it has been moved to the gif plugin).

**Comment by stonebig** [comment]
I did the little goggling earlier, with same lack of example of writing.
For reading, this one example worked (with the patch suggested by @radarhere  : 

https://github.com/python-pillow/Pillow/issues/1525
````
import PIL
import PIL.Image
i = PIL.Image.open('index.gif')
i.show() # correct image displays on left
i.seek(i.tell() + 1)
i.show() # image is incorrect on right
````
![index](https://cloud.githubusercontent.com/assets/4312421/12008015/d6b4bb18-ac21-11e5-8632-283c3d587dbb.gif)

**Comment by WarrenWeckesser** [comment]
Cool, but that *starts* with an animated GIF.  We'll need to be able to create one "from scratch".

I took a look at the Pillow unit tests for the gif plugin, and based on that, here's what I have so far.  The GIF should be a yellow rectangle that fades to black:

```
import numpy as np
from PIL import Image, GifImagePlugin


width = 80
height = 50
num_frames = 10
data = np.zeros((num_frames, height, width, 3), dtype=np.uint8)
for k, val in enumerate(np.linspace(255, 0, num_frames)):
    data[k, :, :, 0:2] = int(val)

im = Image.new('RGB', (height, width))
im = im.convert('P', palette=Image.WEB)
#im = im.convert('P', palette=Image.ADAPTIVE, colors=256)

with open('foo.gif', 'wb') as f:
    for s in GifImagePlugin.getheader(im)[0]:
        f.write(s)
    for a in data:
        imdata = Image.fromarray(a)
        imdata = imdata.convert('P', palette=Image.WEB)
        #imdata = imdata.convert('P', palette=Image.ADAPTIVE, colors=256)
        d = GifImagePlugin.getdata(imdata, duration=250)
        for s in d:
            f.write(s)
    f.write(b';')
    f.close()
```

Using `palette=Image.ADAPTIVE` doesn't work as expected.  I don't know why, but I just started experimenting with this.  `palette=Image.WEB` limits the colors pretty severely.

**Comment by stonebig** [comment]
promising !
![foo](https://cloud.githubusercontent.com/assets/4312421/12008101/691a0df6-ac26-11e5-9560-fc2355518675.gif)

**Comment by WarrenWeckesser** [comment]
I put a "proof of concept" animated GIF writer class in a gist: https://gist.github.com/WarrenWeckesser/7c8423d3935cad422774

Here's an example of its use:

```
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import animation
from animated_gif_writer import AnimatedGifWriter


def update_line(num, data, line):
    line.set_data(data[:, :num+1])
    return line,

fig = plt.figure(figsize=(4.1, 4.0))
ax = fig.add_subplot(111, xlim=(-1, 1), ylim=(-1, 1),
                     autoscale_on=False,  aspect='equal',
                     title="Matplotlib Animation")

num_frames = 8

theta = np.linspace(0, num_frames*np.pi, num_frames)
data = np.exp(1j*theta).view(np.float64).reshape(-1, 2).T

lineplot, = ax.plot([], [], 'b-', linewidth=1.5)
# Reduce the number of colors by turning off antialiasing.
lineplot.set_antialiased(False)

ani = animation.FuncAnimation(fig, update_line, frames=num_frames,
                              init_func=lambda: None,
                              fargs=(data, lineplot))
ani.save('mpl_example.gif', dpi=60, fps=2, writer='pillow-gif')
# Or...
#   writer = AnimatedGifWriter(fps=2)
#   ani.save('mpl_example.gif', dpi=60, writer=writer)
```
It creates:

![mpl_example](https://cloud.githubusercontent.com/assets/321463/12040616/827a2f92-ae38-11e5-807f-6e42afb58955.gif)

**Comment by stonebig** [comment]
cool !

**Comment by WarrenWeckesser** [comment]
@stonebig You might also be interested in animated *PNG* files.  If so, check out `numpngw`: https://pypi.python.org/pypi/numpngw

**Comment by stonebig** [comment]
Even Greater, animated png are much more powerfull ! How to pack that so that these two awesome bleeding edge leaders can be easily switch away from  `writer='imagemagick'` ?
- Jake Vanderplas: https://jakevdp.github.io/blog/2013/05/28/a-simple-animation-the-magic-triangle/
- Holoviews: https://github.com/ioam/holoviews/blob/master/doc/Tutorials/Showcase.ipynb

**Comment by WarrenWeckesser** [comment]
Jake's notebook generates an animated GIF with:

```
anim = animation.FuncAnimation(fig, animate_as_gif,
                               frames=2 * Nframes + 30, interval=50)
anim.save('MagicTriangle.gif', writer='imagemagick')
```

You can generate an animated PNG by adding the import `from numpngw import AnimatedPNGWriter`, and changing `anim.save('MagicTriangle.gif', writer='imagemagick')` to

```
writer = AnimatedPNGWriter(fps=1000.0/anim._interval)
anim.save('MagicTriangle.png', dpi=72, writer=writer)
```
![magictriangle](https://cloud.githubusercontent.com/assets/321463/12043744/f177441c-ae55-11e5-921b-3d59f4eecc0f.png)

**Comment by WeatherGod** [comment]
Weird, in gmail, that animated PNG was not animated when embedded in the
email, but worked fine when clicked and viewed separately... Animated GIFs
work fine no matter what.

**Comment by stonebig** [comment]
hum, gmail is maybe not showing the real image, like blackberry servers tends to do, until you download it  ?

**Comment by stonebig** [comment]
Wheater test (to check for discriminatory practice from google): 
![magictriangle](https://cloud.githubusercontent.com/assets/4312421/12044128/b22d0cc4-ae8c-11e5-974d-02f833336ab3.gif)

**Comment by WeatherGod** [comment]
That one works.

**Comment by stonebig** [comment]
Why Google truncates the smaller "png" one is a mystery. On a side note, the 'animated png' doesn't work on Windows Phone 8 default and unique browser.

**Comment by WarrenWeckesser** [comment]
@stonebig: Animated PNG also doesn't work in Safari.  It works in Firefox, and I've learned that it can work in Chrome and Opera with an appropriate extension installed.  I don't know the status of animated PNG in any Windows browsers (other than what you just reported).

**Comment by WeatherGod** [comment]
Would also be nice to find out how well it works in PowerPoint. Codec
issues are hell at conferences.

**Comment by WarrenWeckesser** [comment]
Earlier I wrote:

> Animated PNG also doesn't work in Safari.

I just saw this at http://animatedpngs.com/:

> Safari is the default browser for Macs. WebKit, Safari's rendering engine, powers all browsers on iOS, meaning iPhones, iPads, and iPods with iOS 8+ now support APNGs.

I have Safari 9.0.2 on my laptop, which doesn't show the animation when displaying an animated PNG, but perhaps newer versions of Safari support it.

**Comment by stonebig** [comment]
Reading the Wikipedia entry about "animated png" (https://en.wikipedia.org/wiki/APNG), I see:
- the consensus has never been reached between "APNG" and "MNG", because of "whatever",
- the situation is blocked since 2007 when the PNG group officially rejected APNG as an official extension and  not [rosy](https://en.wikipedia.org/wiki/Comparison_of_web_browsers#Image_format_support).

Is there some movement to resolve this around a third format that would avoid anyone loosing face ? (it's not clear to me, reading the internet)

**Comment by QuLogic** [comment]
Animated GIF with Pillow is implemented in #10240.
